### PR TITLE
Support the variadic function in FFI downcall

### PR DESCRIPTION
**Pull request by ChengJin01**
The changes aim to support the variadic function with
variadic arguments by setting with ffi_prep_cif_var()
in preparing the ffi specific data for downcall.

Signed-off-by: Cheng Jin <jincheng@ca.ibm.com>

### CODE REVIEW

**Comment by ChengJin01** [comment]
The fix is to support variadic functions in downcall, which was detected in jtreg tests on JDK17/18/19
at
https://github.com/ibmruntimes/openj9-openjdk-jdk17/blob/openj9/test/jdk/java/foreign/TestVarArgs.java
https://github.com/ibmruntimes/openj9-openjdk-jdk18/blob/openj9/test/jdk/java/foreign/TestVarArgs.java
https://github.com/ibmruntimes/openj9-openjdk-jdk19/blob/openj9/test/jdk/java/foreign/TestVarArgs.java

There is no change required in JDK18/19+ except a few code in the layout attributes updated in JDK17 at https://github.com/ibmruntimes/openj9-openjdk-jdk17/pull/112.

**Comment by ChengJin01** [comment]
Reviewer: @tajila
FYI: @gacholio, @pshipton, @DanHeidinga

File: jcl/src/java.base/share/classes/openj9/internal/foreign/abi/LayoutStrPreprocessor.java [hunk 1]
```diff
@@ -49,6 +51,60 @@
...
+		} else if (arch.startsWith("ppc64")) { //$NON-NLS-1$
+			if (osName.startsWith("linux")) { //$NON-NLS-1$
+				VARARGS_ATTR_NAME = "abi/ppc64/sysv/varargs"; //$NON-NLS-1$
+			} else {
+				VARARGS_ATTR_NAME = "abi/ppc64/aix/varargs"; //$NON-NLS-1$
+			}
+		} else if (arch.equals("s390x") && osName.startsWith("linux")) { //$NON-NLS-1$ //$NON-NLS-2$
+			VARARGS_ATTR_NAME = "abi/s390x/sysv/varargs"; //$NON-NLS-1$
+		} else {
+			throw new ExceptionInInitializerError("Unsupported platform: " + arch + "_" + osName); //$NON-NLS-1$ //$NON-NLS-2$
```
Comment (gacholio) [inline jcl/src/java.base/share/classes/openj9/internal/foreign/abi/LayoutStrPreprocessor.java:82]: Do not throw `ExceptionInInitializerError` explicitly - the VM will wrap whatever exception you throw. I suggest simply using `Exception`.

File: runtime/vm/OutOfLineINL_openj9_internal_foreign_abi_InternalDowncallHandler.cpp [hunk 2]
```diff
@@ -157,7 +160,12 @@ OutOfLineINL_openj9_internal_foreign_abi_InternalDowncallHandler_initCifNativeTh
 		goto freeAllMemoryThenExit;
 	}
 
-	if (FFI_OK != ffi_prep_cif(cif, FFI_DEFAULT_ABI, argTypesCount, returnType, &(argTypes[0]))) {
+	/* The variadic argument index is -1 by default if it doesn't exist in the argument list.
+	 * Note: it is literally equal to the count of the fixed arguments before variadic arguments.
+	 */
+	status = (varArgIndex < 0) ? ffi_prep_cif(cif, FFI_DEFAULT_ABI, argTypesCount, returnType, &(argTypes[0]))
```
Comment (gacholio) [inline runtime/vm/OutOfLineINL_openj9_internal_foreign_abi_InternalDowncallHandler.cpp:166]: This would read a lot better as if/else rather than ?.

Comment (ChengJin01) [reply runtime/vm/OutOfLineINL_openj9_internal_foreign_abi_InternalDowncallHandler.cpp]: Agreed and replaced with the if/else statements.

File: jcl/src/java.base/share/classes/openj9/internal/foreign/abi/LayoutStrPreprocessor.java (hunk 1 above)
Comment (ChengJin01) [reply jcl/src/java.base/share/classes/openj9/internal/foreign/abi/LayoutStrPreprocessor.java]: Replaced with `Exception` as suggested above.

Comment (ChengJin01) [reply jcl/src/java.base/share/classes/openj9/internal/foreign/abi/LayoutStrPreprocessor.java]: Changed to `InternalError` to pass the compilation.

File: jcl/src/java.base/share/classes/openj9/internal/foreign/abi/LayoutStrPreprocessor.java [hunk 3]
```diff
@@ -75,8 +131,14 @@ private static int getCountOfPaddingBytesOfStruct(MemoryLayout targetLayout) {
 					long tempPaddingBytes = structElement.byteSize();
 					/* The padding bits must be 8 bits (1 byte), 16 bits (2 bytes), 24 bits (3 bytes)
 					 * or 32 bits (4 bytes) as requested by ffi_call.
+					 *
+					 * Note:
+					 * there are a couple of padding bytes allowed on platforms as follows;
+					 * 1) 5 padding bytes for struct [bool/byte, short, double] (padding between short and double),
```
[the remark this answers is not in the document]
Comment (ChengJin01) [reply jcl/src/java.base/share/classes/openj9/internal/foreign/abi/LayoutStrPreprocessor.java:137]: There is no difference in terms of padding for struct `[short, double]` and `[short, long]` at this point. The 5 padding bytes is intended for `[bool/byte, short, long]` or `[bool/byte, short, double]` as the 5-byte padding is required between `short` and `long or double` on the majority of supported platforms. I just updated the comment for clarification.

File: jcl/src/java.base/share/classes/openj9/internal/foreign/abi/LayoutStrPreprocessor.java [hunk 4]
```diff
@@ -78,9 +134,9 @@ private static int getCountOfPaddingBytesOfStruct(MemoryLayout targetLayout) {
 					 *
 					 * Note:
 					 * there are a couple of padding bytes allowed on platforms as follows;
-					 * 1) 5 padding bytes for struct [bool/byte, short, double] (padding between short and double),
-					 * 2) 6 padding bytes for struct [short, long],
-					 * 3) 7 padding bytes for struct [bool/byte, long].
+					 * 1) 5 padding bytes for struct [bool/byte, short, long/double] (padding between short and long/double),
```
[the remark this answers is not in the document]
Comment (tajila) [reply jcl/src/java.base/share/classes/openj9/internal/foreign/abi/LayoutStrPreprocessor.java:137]: oh I see now, bool/byte + short <-> long/double is 5 bytes

[the remark this answers is not in the document]
Comment (tajila) [reply jcl/src/java.base/share/classes/openj9/internal/foreign/abi/LayoutStrPreprocessor.java:137]: perhaps there should be a `+` in there to make it more clear

**Review by tajila — APPROVED** [review]
just minor comment

File: jcl/src/java.base/share/classes/openj9/internal/foreign/abi/LayoutStrPreprocessor.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (ChengJin01) [reply jcl/src/java.base/share/classes/openj9/internal/foreign/abi/LayoutStrPreprocessor.java]: Replaced `,` with `+` between `bool/byte` and `short` as suggested above.

**Comment by tajila** [comment]
jenkins test sanity win,plinux jdk19

**Pull request merged**